src/main/scala/no/ndla/imageapi/service/WriteService.scala

```diff
@@ -0,0 +1,69 @@
+package no.ndla.imageapi.service
+
+import java.io.ByteArrayInputStream
+import java.lang.Math.max
+
+import com.typesafe.scalalogging.LazyLogging
+import no.ndla.imageapi.model.ValidationException
+import no.ndla.imageapi.model.api.{ImageMetaInformation, NewImageMetaInformation}
+import no.ndla.imageapi.model.domain.Image
+import no.ndla.imageapi.repository.ImageRepository
+import no.ndla.imageapi.service.search.IndexService
+import org.scalatra.servlet.FileItem
+
+import scala.util.{Failure, Random, Success, Try}
+
+trait WriteService {
+  this: ConverterService with ValidationService with ImageRepository with IndexService with ImageStorageService =>
+  val writeService: WriteService
+
+  class WriteService extends LazyLogging {
+    def storeNewImage(newImage: NewImageMetaInformation, file: FileItem): Try[ImageMetaInformation] = {
+      validationService.validateImageFile(file) match {
+        case Some(validationMessage) => return Failure(new ValidationException(errors=Seq(validationMessage)))
+        case _ =>
+      }
+
+      val uploadedImage = uploadImage(file) match {
+        case Failure(e) => return Failure(e)
+        case Success(image) => image
+      }
+
+      val imageMetaInformation = for {
+        imageMeta <- Try(converterService.asDomainImageMetaInformation(newImage, uploadedImage))
+        _ <- validationService.validate(imageMeta)
+        r <- Try(imageRepository.insert(imageMeta))
+        _ <- indexService.indexDocument(r)
+      } yield converterService.asApiImageMetaInformationWithApplicationUrl(r)
+
+      if (imageMetaInformation.isFailure)
+        imageStorage.deleteObject(uploadedImage.fileName)
```

Comment:
Nvm. Ser ut som det er jeg som blander, og at du må blande inn Futures for at det skal bli parallell-eksekvering. I så fall ser det kosher ut...